yt_dlp/networking/impersonate.py

```diff
@@ -0,0 +1,119 @@
+     - `_SUPPORTED_IMPERSONATE_TARGET_TUPLES`: a tuple of supported targets to impersonate.
+        Any Request with an impersonate target not in this list will raise an UnsupportedRequest.
+        Set to None to disable this check.
+     - `_SUPPORTED_IMPERSONATE_TARGET_TUPLE_MAP`: a dict mapping supported targets to custom targets.
+        This works similar to `_SUPPORTED_IMPERSONATE_TARGET_TUPLES`.
+
+    Note: Only one of `_SUPPORTED_IMPERSONATE_TARGET_TUPLE_MAP` and `_SUPPORTED_IMPERSONATE_TARGET_TUPLES` can be defined.
+    Note: Entries are in order of preference
+
+    Parameters:
+    @param impersonate: the default impersonate target to use for requests.
+                        Set to None to disable impersonation.
+    """
+    _SUPPORTED_IMPERSONATE_TARGET_TUPLES: tuple[ImpersonateTarget] = ()
+    _SUPPORTED_IMPERSONATE_TARGET_TUPLE_MAP: dict[ImpersonateTarget, Any] = {}
+
+    def __init__(self, *, impersonate: ImpersonateTarget = None, **kwargs):
+        super().__init__(**kwargs)
+        self.impersonate = impersonate
+
+    def _check_impersonate_target(self, target: ImpersonateTarget):
+        assert isinstance(target, (tuple, NoneType))
+        if target is None or not self.get_supported_targets():
+            return
+        if not self.is_supported_target(target):
+            raise UnsupportedRequest(f'Unsupported impersonate target: {target}')
+
+    def _check_extensions(self, extensions):
+        super()._check_extensions(extensions)
+        if 'impersonate' in extensions:
+            self._check_impersonate_target(extensions.get('impersonate'))
+
+    def _validate(self, request):
+        super()._validate(request)
+        self._check_impersonate_target(self.impersonate)
+
+    def _resolve_target(self, target: ImpersonateTarget | None):
+        """Resolve a target to a supported target."""
+        if target is None:
+            return
+        for supported_target in self.get_supported_targets():
+            if _target_within(target, supported_target):
+                if self.verbose:
+                    self._logger.stdout(
+                        f'{self.RH_NAME}: resolved impersonate target {target} to {supported_target}')
+                return supported_target
+
+    def get_supported_targets(self) -> tuple[ImpersonateTarget]:
+        return tuple(self._SUPPORTED_IMPERSONATE_TARGET_TUPLE_MAP.keys()) or tuple(self._SUPPORTED_IMPERSONATE_TARGET_TUPLES)
+
+    def is_supported_target(self, target: ImpersonateTarget):
+        return self._resolve_target(target) is not None
+
+    def _get_request_target(self, request):
+        """Get the requested target for the request"""
+        return request.extensions.get('impersonate') or self.impersonate
+
+    def _get_resolved_request_target(self, request) -> ImpersonateTarget:
+        """Get the resolved target for this request. This gives the matching supported target"""
+        return self._resolve_target(self._get_request_target(request))
```

Comment:
Unresolved target is useless, no?

```suggestion
    def _get_request_target(self, request) -> ImpersonateTarget:
        """Get the resolved requested target for the request"""
        return _resolve_target(request.extensions.get('impersonate') or self.impersonate)
```